library/src/armory/metrics/detection.py

```diff
@@ -0,0 +1,189 @@
+        iou_threshold: float = 0.5,
+        score_threshold: float = 0.5,
+        class_list: Optional[Sequence[int]] = None,
+    ):
+        """
+        Initialize the metric.
+
+        Args:
+            mean: `True` to return the mean values across all images or `False`
+                to return the individual values per image
+            iou_threshold: The minimum IoU for detected boxes to be considered
+                and evaluated against the target boxes
+            score_threshold: The minimum prediction score for boxes to be
+                considered and evaluated against the target boxes
+            class_list: Optional list of classes, such that all predictions with
+                labels and ground-truths with labels _not_ in the class list are
+                to be ignored
+        """
+        super().__init__()
+        self.mean = mean
+        self.iou_threshold = iou_threshold
+        self.score_threshold = score_threshold
+        self.class_list = (
+            torch.as_tensor(class_list) if class_list is not None else None
+        )
+
+        self.true_positive_rate_per_img: List[torch.Tensor]
+        self.misclassification_rate_per_img: List[torch.Tensor]
+        self.disappearance_rate_per_img: List[torch.Tensor]
+        self.hallucinations_per_img: List[torch.Tensor]
+
+        self.add_state("true_positive_rate_per_img", default=[], dist_reduce_fx="cat")
+        self.add_state(
+            "misclassification_rate_per_img", default=[], dist_reduce_fx="cat"
+        )
+        self.add_state("disappearance_rate_per_img", default=[], dist_reduce_fx="cat")
+        self.add_state("hallucinations_per_img", default=[], dist_reduce_fx="cat")
+
+    def update(
+        self,
+        preds: Sequence[BoundingBoxes.BoxesTorch],
+        targets: Sequence[BoundingBoxes.BoxesTorch],
+    ):
+        for y, y_pred in zip(targets, preds):
+            if self.class_list:
+                # Filter out ground-truth classes with labels not in class_list
+                indices_to_keep = torch.where(torch.isin(y["labels"], self.class_list))
+                gt_boxes = y["boxes"][indices_to_keep]
+                gt_labels = y["labels"][indices_to_keep]
+            else:
+                gt_boxes = y["boxes"]
+                gt_labels = y["labels"]
+
+            # initialize count of hallucinations
+            num_hallucinations = torch.tensor(0)
+            num_gt_boxes = len(gt_boxes)
+
+            # Initialize arrays that will indicate whether each respective ground-truth
+            # box is a true positive or misclassified
+            true_positive_array = torch.zeros((num_gt_boxes,))
+            misclassification_array = torch.zeros((num_gt_boxes,))
+
+            # Only consider the model's confident predictions
+            assert y_pred["scores"] is not None
+            conf_pred_indices = torch.where(y_pred["scores"] > self.score_threshold)[0]
+            if self.class_list:
```

Comment:
This test fails if the class list was specified as `class_list=(0,)` in the constructor. Since object detectors assign class 0 to background by convention, this edge case might not make sense.